File: Source/Engine/Graphics/Vulkan/VulkanCommandQueue.cpp
// Copyright © Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

#include "VulkanCommandQueue.h"
#include "VulkanCommandBuffer.h"
#include "VulkanSwapChain.h"
#include "VulkanGraphics.h"

namespace Alimer
{
	VulkanCommandQueue::VulkanCommandQueue(VulkanGraphics& device_, CommandQueueType type, uint32_t queueFamilyIndex, uint32_t queueIndex)
		: CommandQueue(type)
		, device(device_)
	{
		vkGetDeviceQueue(device.GetHandle(), queueFamilyIndex, queueIndex, &handle);

		VkCommandPoolCreateInfo poolInfo{ VK_STRUCTURE_TYPE_COMMAND_POOL_CREATE_INFO };
		poolInfo.flags = VK_COMMAND_POOL_CREATE_TRANSIENT_BIT;
		poolInfo.queueFamilyIndex = queueFamilyIndex;

		for (uint32_t i = 0; i < kMaxFramesInFlight; i++)
		{
			VK_CHECK(vkCreateCommandPool(device.GetHandle(), &poolInfo, nullptr, &commandBuffers[i].commandPool));
		}
	}

	VulkanCommandQueue::~VulkanCommandQueue()
	{
		for (uint32_t i = 0; i < kMaxFramesInFlight; i++)
		{
			for (uint32_t j = 0; j < kMaxCommandLists; j++)
			{
				commandBuffers[i].commandBuffers[j].reset();
			}

			vkDestroyCommandPool(device.GetHandle(), commandBuffers[i].commandPool, nullptr);
		}
	}

	void VulkanCommandQueue::Reset(uint32_t frameIndex_)
	{
		frameIndex = frameIndex_;

		commandBuffers[frameIndex].allocatedBuffers = 0;
		VK_CHECK(vkResetCommandPool(device.GetHandle(), commandBuffers[frameIndex].commandPool, 0));
	}

	void VulkanCommandQueue::AddWaitSemaphore(VkSemaphore semaphore, VkPipelineStageFlags waitDstStageMask)
	{
		ALIMER_ASSERT(semaphore != VK_NULL_HANDLE);

		waitSemaphores.push_back(semaphore);
		waitDstStageMasks.push_back(waitDstStageMask);
        waitValues.push_back(0);
	}

	void VulkanCommandQueue::AddSignalSemaphore(VkSemaphore semaphore)
	{
		ALIMER_ASSERT(semaphore != VK_NULL_HANDLE);

		signalSemaphores.push_back(semaphore);
	}

	void VulkanCommandQueue::QueuePresent(VulkanSwapChain* swapChain)
	{
		ALIMER_ASSERT(queueType == CommandQueueType::Graphics);

		presentSwapChains[presentSwapChainsCount] = swapChain;
		vkPresentSwapChains[presentSwapChainsCount] = swapChain->GetHandle();
		imageIndices[presentSwapChainsCount] = swapChain->GetBackBufferIndex();
		presentSwapChainsCount++;
	}

	CommandBuffer* VulkanCommandQueue::GetCommandBuffer()
	{
		auto& frameCmd = commandBuffers[frameIndex];

		if (frameCmd.allocatedBuffers >= kMaxCommandLists)
		{
            LOGE("Vulkan: Max command buffer count reached!");
			return nullptr;
		}

		if (frameCmd.commandBuffers[frameCmd.allocatedBuffers] == nullptr)
		{
			frameCmd.commandBuffers[frameCmd.allocatedBuffers].reset(new VulkanCommandBuffer(*this, frameCmd.commandPool));
		}

		// Begin recording
		frameCmd.commandBuffers[frameCmd.allocatedBuffers]->Reset(frameIndex);

		return frameCmd.commandBuffers[frameCmd.allocatedBuffers++].get();
	}

	void VulkanCommandQueue::WaitIdle()
	{
		VK_CHECK(vkQueueWaitIdle(handle));
	}

	void VulkanCommandQueue::Submit(CommandBuffer* const* commandBuffers, uint32_t count, bool waitForCompletion)
	{
        const uint64_t copyFenceValue = device.FlushCopy();

        // Sync with copy queue
        if (copyFenceValue > 0)
        {
            waitDstStageMasks.push_back(VK_PIPELINE_STAGE_TRANSFER_BIT);
            waitSemaphores.push_back(device.GetCopySemaphore());
            waitValues.push_back(copyFenceValue);
        }

		std::vector<VkCommandBuffer> vkCommandBuffers;

		for (uint32_t i = 0; i < count; i++)
		{
			VulkanCommandBuffer* vulkanCommandBuffer = ToVulkan(commandBuffers[i]);
			vulkanCommandBuffer->End();
			vkCommandBuffers.push_back(vulkanCommandBuffer->GetHandle());
		}

		VkSubmitInfo submitInfo{ VK_STRUCTURE_TYPE_SUBMIT_INFO };
		submitInfo.waitSemaphoreCount = static_cast<uint32_t>(waitSemaphores.size());
		submitInfo.pWaitSemaphores = waitSemaphores.data();
		submitInfo.pWaitDstStageMask = waitDstStageMasks.data();
		submitInfo.commandBufferCount = count;
		submitInfo.pCommandBuffers = vkCommandBuffers.data();
		submitInfo.signalSemaphoreCount = static_cast<uint32_t>(signalSemaphores.size());
		submitInfo.pSignalSemaphores = signalSemaphores.data();

        VkTimelineSemaphoreSubmitInfo timelineSubmitInfo{ VK_STRUCTURE_TYPE_TIMELINE_SEMAPHORE_SUBMIT_INFO };
        timelineSubmitInfo.pNext = nullptr;
        if (waitValues.size())
        {
            timelineSubmitInfo.waitSemaphoreValueCount = (uint32_t)waitValues.size();
            timelineSubmitInfo.pWaitSemaphoreValues = waitValues.data();
            timelineSubmitInfo.signalSemaphoreValueCount = (uint32_t)signalValues.size();
            timelineSubmitInfo.pSignalSemaphoreValues = signalValues.data();

            submitInfo.pNext = &timelineSubmitInfo;
        }

		VkFence fence = device.AcquireFence();

		VkResult result = vkQueueSubmit(handle, 1, &submitInfo, fence);
		if (result != VK_SUCCESS)
		{
            LOGE("Vulkan: Failed to submit command buffer!");
			return;
		}

		if (waitForCompletion)
		{
			vkWaitForFences(device.GetHandle(), 1, &fence, VK_TRUE, 100000000000);
			device.ReleaseFence(fence);
		}
		else
		{
			device.SubmitFence(fence);
		}

		// Handle automatic SwapChain queue present
		if (queueType == CommandQueueType::Graphics)
		{
			if (presentSwapChainsCount > 0)
			{
				VkPresentInfoKHR presentInfo{ VK_STRUCTURE_TYPE_PRESENT_INFO_KHR };
				presentInfo.waitSemaphoreCount = static_cast<uint32_t>(signalSemaphores.size());
				presentInfo.pWaitSemaphores = signalSemaphores.data();
				presentInfo.swapchainCount = presentSwapChainsCount;
				presentInfo.pSwapchains = vkPresentSwapChains.data();
				presentInfo.pImageIndices = imageIndices.data();
				presentInfo.pResults = presentResults.data();

				vkQueuePresentKHR(handle, &presentInfo);

				for (size_t i = 0; i < presentSwapChainsCount; i++)
				{
					presentSwapChains[i]->AfterPresent(presentResults[i]);
				}
			}
		}

        // Clear wait and signal data
        waitSemaphores.clear();
        waitDstStageMasks.clear();
        signalSemaphores.clear();
        waitValues.clear();
        signalValues.clear();
        presentSwapChainsCount = 0;
	}
}

File: src/Alimer/Graphics/Vulkan/VulkanCommandQueue.cs
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

using Vortice.Vulkan;
using static Vortice.Vulkan.Vulkan;
using static Alimer.Graphics.Constants;
using System.Diagnostics;

namespace Alimer.Graphics.Vulkan;

internal unsafe class VulkanCommandQueue : CommandQueue, IDisposable
{
    public readonly VkQueue Handle;
    public readonly object LockObject = new();
    private readonly VkSemaphore _semaphore = VkSemaphore.Null;
    private readonly VkFence[] _frameFences;

    private uint _commandBufferCount = 0;
    private readonly List<VulkanCommandBuffer> _commandBuffers = [];
    private readonly List<VkCommandBuffer> _submitCommandBuffers = [];

    private readonly List<VulkanSwapChain> _presentSwapChains = [];
    private readonly List<VkSemaphore> _submitSignalSemaphores = [];

    public VulkanCommandQueue(VulkanGraphicsDevice device, CommandQueueType queueType)
    {
        VkDevice = device;
        QueueType = queueType;

        uint queueFamilyIndex = device.GetQueueFamily(queueType);
        uint queueIndex = device.GetQueueIndex(queueType);
        device.DeviceApi.vkGetDeviceQueue(device.Handle, queueFamilyIndex, queueIndex, out Handle);

        VkSemaphoreTypeCreateInfo timelineCreateInfo = new()
        {
            pNext = null,
            semaphoreType = VkSemaphoreType.Timeline,
            initialValue = 0
        };

        VkSemaphoreCreateInfo createInfo = new()
        {
            pNext = &timelineCreateInfo,
            flags = 0
        };

        device.DeviceApi.vkCreateSemaphore(device.Handle, &createInfo, null, out _semaphore).CheckResult();

        _frameFences = new VkFence[device.MaxFramesInFlight];
        for (int i = 0; i < device.MaxFramesInFlight; ++i)
        {
            device.DeviceApi.vkCreateFence(device.Handle, out _frameFences[i]);
        }
    }

    /// <inheritdoc />
    public override GraphicsDevice Device => VkDevice;

    /// <inheritdoc />
    public override CommandQueueType QueueType { get; }

    public VulkanGraphicsDevice VkDevice { get; }
    public VkFence FrameFence => _frameFences[Device.FrameIndex];

    public void Dispose()
    {
        WaitIdle();
        VkDevice.DeviceApi.vkDestroySemaphore(VkDevice.Handle, _semaphore);

        for (int i = 0; i < _frameFences.Length; ++i)
        {
            VkDevice.DeviceApi.vkDestroyFence(VkDevice.Handle, _frameFences[i]);
        }

        foreach (VulkanCommandBuffer commandBuffer in _commandBuffers)
        {
            commandBuffer.Destroy();
        }
        _commandBuffers.Clear();
    }

    public void FinishFrame()
    {
        _commandBufferCount = 0;
        _submitCommandBuffers.Clear();
        _presentSwapChains.Clear();
    }

    public override void Execute(IEnumerable<CommandBuffer> commandBuffers, bool waitForCompletion = false)
    {
        foreach (VulkanCommandBuffer commandBuffer in commandBuffers)
        {
            VkCommandBuffer handle = commandBuffer.End();
            _submitCommandBuffers.Add(handle);
        }
    }

    /// <inheritdoc />
    public override void WaitIdle()
    {
        VkResult result = VkDevice.DeviceApi.vkQueueWaitIdle(Handle);
        if (result != VK_SUCCESS)
        {
            throw new GraphicsException("Vulkan: Failed to wait for Vulkan queue idle");
        }
    }

    public void QueuePresent(VulkanSwapChain swapChain)
    {
        _presentSwapChains.Add(swapChain);
        _submitSignalSemaphores.Add(swapChain.ReleaseSemaphore);
    }

    public void Submit(VkFence fence)
    {
        if (Handle.IsNull)
            return;

        lock (LockObject)
        {
            uint waitSemaphoreInfoCount = (uint)_presentSwapChains.Count;
            uint signalSemaphoreInfoCount = (uint)_presentSwapChains.Count;
            VkSemaphoreSubmitInfo* waitSemaphoreInfos = stackalloc VkSemaphoreSubmitInfo[_presentSwapChains.Count];
            VkSemaphoreSubmitInfo* signalSemaphoreInfos = stackalloc VkSemaphoreSubmitInfo[_presentSwapChains.Count];

            for (int i = 0; i < _presentSwapChains.Count; i++)
            {
                waitSemaphoreInfos[i] = new()
                {
                    semaphore = _presentSwapChains[i].AcquireSemaphore,
                    value = 0, // not a timeline semaphore
                    stageMask = VkPipelineStageFlags2.ColorAttachmentOutput
                };

                signalSemaphoreInfos[i] = new()
                {
                    semaphore = _presentSwapChains[i].ReleaseSemaphore,
                    value = 0, // not a timeline semaphore
                };

                // Advance surface frame index
                _presentSwapChains[i].AdvanceFrame();
            }

            uint commandBufferInfoCount = (uint)_submitCommandBuffers.Count;
            VkCommandBufferSubmitInfo* commandBufferInfos = stackalloc VkCommandBufferSubmitInfo[_submitCommandBuffers.Count];
            for (int i = 0; i < _submitCommandBuffers.Count; i++)
            {
                commandBufferInfos[i] = new()
                {
                    commandBuffer = _submitCommandBuffers[i]
                };
            }

            VkSubmitInfo2 submitInfo = new()
            {
                waitSemaphoreInfoCount = waitSemaphoreInfoCount,
                pWaitSemaphoreInfos = waitSemaphoreInfos,
                commandBufferInfoCount = commandBufferInfoCount,
                pCommandBufferInfos = commandBufferInfos,
                signalSemaphoreInfoCount = signalSemaphoreInfoCount,
                pSignalSemaphoreInfos = signalSemaphoreInfos
            };

            VkDevice.DeviceApi.vkQueueSubmit2(Handle, 1, &submitInfo, fence).CheckResult();

            if (_presentSwapChains.Count > 0)
            {
                VkSemaphore* pWaitSemaphores = stackalloc VkSemaphore[_presentSwapChains.Count];
                VkSwapchainKHR* pSwapchains = stackalloc VkSwapchainKHR[_presentSwapChains.Count];
                uint* pImageIndices = stackalloc uint[_presentSwapChains.Count];

                for (int i = 0; i < _presentSwapChains.Count; i++)
                {
                    pWaitSemaphores[i] = _submitSignalSemaphores[i];
                    pSwapchains[i] = _presentSwapChains[i].Handle;
                    pImageIndices[i] = _presentSwapChains[i].ImageIndex;
                }

                VkPresentInfoKHR presentInfo = new()
                {
                    waitSemaphoreCount = (uint)_submitSignalSemaphores.Count,
                    pWaitSemaphores = pWaitSemaphores,
                    swapchainCount = (uint)_presentSwapChains.Count,
                    pSwapchains = pSwapchains,
                    pImageIndices = pImageIndices
                };

                VkResult result = VkDevice.DeviceApi.vkQueuePresentKHR(Handle, &presentInfo);
                if (result != VkResult.Success)
                {
                    // Handle outdated error in present
                    if (result == VkResult.SuboptimalKHR || result == VkResult.ErrorOutOfDateKHR)
                    {
                        //for (auto & swapchain : swapchain_updates)
                        //{
                        //    auto internal_state = to_internal(&swapchain);
                        //    bool success = CreateSwapChainInternal(internal_state, device->physicalDevice, device->device, device->allocationhandler);
                        //    assert(success);
                        //}
                    }
                    else
                    {
                        Debug.Assert(false);
                    }
                }
            }

            //swapchain_updates.clear();
            _presentSwapChains.Clear();
            //submit_waitSemaphoreInfos.clear();
            _submitSignalSemaphores.Clear();
            //submit_signalSemaphoreInfos.clear();
            //submit_cmds.clear();
        }
    }

    public CommandBuffer AcquireCommandBuffer(Utf8ReadOnlyString label = default)
    {
        uint index = _commandBufferCount++;
        VulkanCommandBuffer commandBuffer;
        if (index >= _commandBuffers.Count)
        {
            commandBuffer = new VulkanCommandBuffer(this);
            _commandBuffers.Add(commandBuffer);
        }
        else
        {
            commandBuffer = _commandBuffers[_commandBuffers.Count - 1];
        }

        commandBuffer.Begin(Device.FrameIndex, label);
        return commandBuffer;
    }
}
